reference:
  CATCH_EXCEPTIONS false

object Harness

contract list
  Execute


Instantiate task expects 'Test'
  set 
    @pass to $pass
    @fail to $fail
    @ids to $ids
    @SAI to $SAI


Log task 
  if @ids[$]
    debug 'Duplicate test: "${$}"'
  set @ids[$] to true



Test task as description, function
  set description to '${@isa}: ${self}'
  @Log description
  
  if CATCH_EXCEPTIONS
    set success to false
    try 
      set success from function
    catch 
      debug error
      @fail.push err
  else
    set success from function

  if success
    @pass.push description
  else
    @fail.push 'test "${description}" failed:\n${function.toString()}'



Match task as description, expected, function
  set description to '${@isa}: ${self}'
  @Log description
  
  if CATCH_EXCEPTIONS
    try 
      set result from function
    catch 
      debug error
      set result to error
      @fail.push error
  else
    set result from function
    
  if result = expected
    @pass.push description
  else
    @fail.push 'test "${description}" expected ${expected}, got ${result} in:\n${function.toString()}'


